File: scala-advent-of-code/src/main/scala/advent/twenty_fifteen/Day07SomeAssemblyRequired.scala
package advent.twenty_fifteen

import advent.common.DailyProblem
import advent.twenty_fifteen.support._
import advent.utilities.FileUtilities

import scala.annotation.tailrec

class Day07SomeAssemblyRequired(filename: String) extends DailyProblem[Map[String, Int], Map[String, Int]] {
  private val signalPattern = "([a-z0-9]+) -> ([a-z]+)".r
  private val andPattern = "([a-z0-9]+) AND ([a-z]+) -> ([a-z]+)".r
  private val orPattern = "([a-z0-9]+) OR ([a-z]+) -> ([a-z]+)".r
  private val lshiftPattern = "([a-z]+) LSHIFT ([0-9]+) -> ([a-z]+)".r
  private val rshiftPattern = "([a-z]+) RSHIFT ([0-9]+) -> ([a-z]+)".r
  private val notPattern = "NOT ([a-z]+) -> ([a-z]+)".r

  private val wirePattern = "([a-z]+)".r
  private val constantPattern = "([0-9]+)".r

  private def parse(value: String): Input = {
    value match {
      case wirePattern(name) => WireInput(name)
      case constantPattern(constant) => ConstantInput(constant.toInt)
      case _ => throw new IllegalStateException("Cannot match '" + value + "'")
    }
  }

  private def parser(line: String): LogicOperation =
    line match {
      case signalPattern(in, out) => SignalOperation(parse(in), out)
      case andPattern(lhs, rhs, out) => AndOperation(parse(lhs), parse(rhs), out)
      case orPattern(lhs, rhs, out) => OrOperation(parse(lhs), parse(rhs), out)
      case lshiftPattern(in, count, out) => LShiftOperation(parse(in), count.toInt, out)
      case rshiftPattern(in, count, out) => RShiftOperation(parse(in), count.toInt, out)
      case notPattern(in, out) => NotOperation(parse(in), out)
      case _ => throw new IllegalStateException("Cannot match '" + line + "'")
    }

  private val operations = FileUtilities.readFile(filename, parser)

  @tailrec
  private def evaluate(operations: List[LogicOperation], pending: List[LogicOperation], wires: Map[String, Int]): Map[String, Int] = {
    if (operations.isEmpty && pending.isEmpty) {
      wires
    } else if (operations.isEmpty && pending.nonEmpty) {
      evaluate(pending, List(), wires)
    } else if (operations.head.ready(wires)) {
      evaluate(operations.tail, pending, operations.head.update(wires))
    } else {
      evaluate(operations.tail, operations.head :: pending, wires)
    }
  }

  override val part1Answer: Map[String, Int] = evaluate(operations, List(), Map())
  override val part2Answer: Map[String, Int] = evaluate(operations.filter(operation => operation match {
    case SignalOperation(_, "b") => false
    case _ => true
  }), List(), Map("b" -> part1Answer.getOrElse("a", 0)))
}

File: java-advent-of-code/src/main/java/advent/twenty_fifteen/Day07SomeAssemblyRequired.java
package advent.twenty_fifteen;

import advent.common.DailyProblem;
import advent.twenty_fifteen.support.BitwiseParser;
import advent.twenty_fifteen.support.IBitwiseLogic;
import advent.utilities.FileUtilities;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class Day07SomeAssemblyRequired implements DailyProblem<Map<String, Integer>, Map<String, Integer>> {
    private final Map<String, Integer> part1Answer;
    private final Map<String, Integer> part2Answer;

    public Day07SomeAssemblyRequired(String filename) {
        List<IBitwiseLogic> operations = FileUtilities.readLines(filename, BitwiseParser::PARSE_BITWISE);

        this.part1Answer = new HashMap<>();
        executeLogic(part1Answer, new LinkedList<>(operations));

        this.part2Answer = new HashMap<>();
        part2Answer.put("b", part1Answer.get("a"));
        executeLogic(part2Answer, new LinkedList<>(operations));
    }

    private static void executeLogic(Map<String, Integer> values, List<IBitwiseLogic> working) {
        boolean somethingChanged = true;
        while (somethingChanged) {
            int previous = values.size();

            for (IBitwiseLogic operation : new LinkedList<>(working)) {
                if (operation.canBeApplied(values)) {
                    operation.apply(values);
                    working.remove(operation);
                }
            }
            somethingChanged = values.size() > previous;
        }
    }

    @Override
    public Map<String, Integer> getPart1Answer() {
        return part1Answer;
    }

    @Override
    public Map<String, Integer> getPart2Answer() {
        return part2Answer;
    }
}
